hibernate-core/src/main/java/org/hibernate/event/service/internal/EventListenerRegistryImpl.java

```diff
@@ -149,7 +151,17 @@ public void prepare(MetadataImplementor metadata) {
 					propertyIterator.hasNext(); ) {
 				Property property = (Property) propertyIterator.next();
 
-				if ( property.getType().isComponentType() ) {
+				boolean mustBuildCallbacksForEmbeddable;
+				if(property.getType().isCollectionType()){
+					CollectionType collectionType = (CollectionType) property.getType();
+					final Type collectionElementType = collectionType.getElementType(getSessionFactory());
+					mustBuildCallbacksForEmbeddable = collectionElementType.isComponentType();
+				}
+				else {
+					mustBuildCallbacksForEmbeddable = property.getType().isComponentType();
+				}
```

Comment:
There seems to be some room for code logic improvement. For instance
```
Type type = property.getType();
if ( type.isCollectionType() ) {
    type = ( ( CollectionType ) type ).getElementType( getSessionFactory() );
}

if ( type.isComponentType() ) { // instead of 'mustBuildCallbackForEmbeddable'
    ...
}
```

Also, it is highly recommended to use empty char delimiter consistently.
Feel free to ignore my comment. As usual, I am notorious to suggest generic coding improvement without approving for lack of enough domain knowledge, :).